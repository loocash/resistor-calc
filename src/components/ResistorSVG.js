import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

function ResistorSVG({ code }) {

  const bands = code.length;

  if (!_.inRange(bands, 3, 6)) {
    throw new Error(`ResistorSVG: expected array of 3, 4 or 5 elements, got: ${bands}`);
  }

  code = code.map(color => color.toLowerCase());

  const width = (bands === 5) ? 35 : 45;

  const bandsArr = {
    3: [
      {
        x: 185.05522,
        y: 193.97354,
        height: 150.48511,
        width,
      },
      {
        x: 226.65579,
        y: 18.901487,
        height: 112.64787,
        width,
      },
      {
        x: 338.36636,
        y: 18.901487,
        height: 112.64787,
        width,
      },
    ],
    4: [
      {
        x: 185.05522,
        y: 193.97354,
        height: 150.48511,
        width,
      },
      {
        x: 226.65579,
        y: 18.901487,
        height: 112.64787,
        width,
      },
      {
        x: 338.36636,
        y: 18.901487,
        height: 112.64787,
        width,
      },
      {
        x: 447.96912,
        y: 0.0264872,
        height: 150.48511,
        width,
      },
    ],
    5: [
      {
        x: 160,
        y: 193.97354,
        height: 150.48511,
        width,
      },
      {
        x: 155,
        y: 0,
        height: 150.48511,
        width,
      },
      {
        x: 245,
        y: 18.901487,
        height: 112.64787,
        width,
      },
      {
        x: 330,
        y: 18.901487,
        height: 112.64787,
        width,
      },
      {
        x: 447.96912,
        y: 0.0264872,
        height: 150.48511,
        width,
      },
    ],
  };

  const rects = code.map((color, index) =>
    <rect ry="0"
      key={index}
      {...bandsArr[bands][index]}
      className={`is-${color}`}
    />
  );

  return (

    <svg
      className="resistor-svg"
      xmlns="http://www.w3.org/2000/svg"
      height="42.473526mm"
      width="174.7043mm"
      viewBox="0 0 619.03099 150.49675"
      id="svg2">

      <rect
        style={{fill: '#808080', fillOpacity: 1}}
        id="rect3292"
        width="100.33115"
        height="27.869762"
        x="522.83673"
        y="-764169.5"
        ry="13.934881" />
      <g
        id="g3298"
        transform="translate(-68.002197,-193.97351)">
        <rect
          style={{fill: '#808080', fillOpacity: 1}}
          id="rect3296"
          width="619.03101"
          height="27.869762"
          x="68.002197"
          y="254.49246"
          ry="13.934881" />
        <g
          style={{fill: '#ffd189', fillOpacity:1}}
          transform="translate(-15.607067,3.3443722)"
          id="g3284">
          <g
            style={{fill: '#ffd189', fillOpacity:1}}
            transform="translate(189.51438,3.3443714)"
            id="g3272">
            <rect
              style={{fill: '#ffd189', fillOpacity:1}}
              id="rect3274"
              width="162.7594"
              height="112.59386"
              x="152.72629"
              y="206.23622"
              ry="0" />
            <rect
              ry="30.099342"
              y="187.28477"
              x="294.30466"
              height="150.49673"
              width="153.84108"
              id="rect3276"
              style={{fill: '#ffd189', fillOpacity:1}} />
          </g>
          <g
            style={{fill: '#ffd189', fillOpacity:1}}
            id="g3278"
            transform="matrix(-1,0,0,-1,595.29807,528.41065)">
            <rect
              ry="0"
              y="206.23622"
              x="152.72629"
              height="112.59386"
              width="162.7594"
              id="rect3280"
              style={{fill: '#ffd189', fillOpacity:1}} />
            <rect
              style={{fill: '#ffd189', fillOpacity:1}}
              id="rect3282"
              width="153.84108"
              height="150.49673"
              x="294.30466"
              y="187.28477"
              ry="30.099342" />
          </g>
        </g>
      </g>
      <g
        transform="translate(-68.002197,-193.97351)">
        { rects[0] }
      </g>
      { rects.slice(1) }
    </svg>

  );
}

ResistorSVG.propTypes = {
  code: PropTypes.array.isRequired,
};

export default ResistorSVG;
