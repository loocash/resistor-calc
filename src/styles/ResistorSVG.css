
.resistor-svg {
    width: 20em;
    height: 30%;
    display: block;
    margin: 0 auto;
}
