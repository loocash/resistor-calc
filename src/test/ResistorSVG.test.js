import React from 'react';

import {render, cleanup} from '@testing-library/react';
import ResistorSVG from '../components/ResistorSVG';

beforeEach(cleanup);

const TestCodes = [
  {
    code: ["Green", "Blue", "Yellow", "Gold"],
  },
  {
    code: ["Red", "Orange", "Violet", "Black", "Brown"],
  }
];

TestCodes.forEach(({code}) => {
  test(`matches snapshot for ${code}`, () => {
    const { container } = render(<ResistorSVG code={code} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
