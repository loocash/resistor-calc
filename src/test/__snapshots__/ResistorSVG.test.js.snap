// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for Green,Blue,Yellow,Gold 1`] = `
<svg
  class="resistor"
  height="42.473526mm"
  id="svg2"
  viewBox="0 0 619.03099 150.49675"
  width="174.7043mm"
  xmlns="http://www.w3.org/2000/svg"
>
  <rect
    height="27.869762"
    id="rect3292"
    ry="13.934881"
    style="fill: #808080; fill-opacity: 1;"
    width="100.33115"
    x="522.83673"
    y="-764169.5"
  />
  <g
    id="g3298"
    transform="translate(-68.002197,-193.97351)"
  >
    <rect
      height="27.869762"
      id="rect3296"
      ry="13.934881"
      style="fill: #808080; fill-opacity: 1;"
      width="619.03101"
      x="68.002197"
      y="254.49246"
    />
    <g
      id="g3284"
      style="fill: #ffd189; fill-opacity: 1;"
      transform="translate(-15.607067,3.3443722)"
    >
      <g
        id="g3272"
        style="fill: #ffd189; fill-opacity: 1;"
        transform="translate(189.51438,3.3443714)"
      >
        <rect
          height="112.59386"
          id="rect3274"
          ry="0"
          style="fill: #ffd189; fill-opacity: 1;"
          width="162.7594"
          x="152.72629"
          y="206.23622"
        />
        <rect
          height="150.49673"
          id="rect3276"
          ry="30.099342"
          style="fill: #ffd189; fill-opacity: 1;"
          width="153.84108"
          x="294.30466"
          y="187.28477"
        />
      </g>
      <g
        id="g3278"
        style="fill: #ffd189; fill-opacity: 1;"
        transform="matrix(-1,0,0,-1,595.29807,528.41065)"
      >
        <rect
          height="112.59386"
          id="rect3280"
          ry="0"
          style="fill: #ffd189; fill-opacity: 1;"
          width="162.7594"
          x="152.72629"
          y="206.23622"
        />
        <rect
          height="150.49673"
          id="rect3282"
          ry="30.099342"
          style="fill: #ffd189; fill-opacity: 1;"
          width="153.84108"
          x="294.30466"
          y="187.28477"
        />
      </g>
    </g>
  </g>
  <g
    transform="translate(-68.002197,-193.97351)"
  >
    <rect
      class="is-green"
      height="150.48511"
      ry="0"
      width="45"
      x="185.05522"
      y="193.97354"
    />
  </g>
  <rect
    class="is-blue"
    height="112.64787"
    ry="0"
    width="45"
    x="226.65579"
    y="18.901487"
  />
  <rect
    class="is-yellow"
    height="112.64787"
    ry="0"
    width="45"
    x="338.36636"
    y="18.901487"
  />
  <rect
    class="is-gold"
    height="150.48511"
    ry="0"
    width="45"
    x="447.96912"
    y="0.0264872"
  />
</svg>
`;

exports[`matches snapshot for Red,Orange,Violet,Black,Brown 1`] = `
<svg
  class="resistor"
  height="42.473526mm"
  id="svg2"
  viewBox="0 0 619.03099 150.49675"
  width="174.7043mm"
  xmlns="http://www.w3.org/2000/svg"
>
  <rect
    height="27.869762"
    id="rect3292"
    ry="13.934881"
    style="fill: #808080; fill-opacity: 1;"
    width="100.33115"
    x="522.83673"
    y="-764169.5"
  />
  <g
    id="g3298"
    transform="translate(-68.002197,-193.97351)"
  >
    <rect
      height="27.869762"
      id="rect3296"
      ry="13.934881"
      style="fill: #808080; fill-opacity: 1;"
      width="619.03101"
      x="68.002197"
      y="254.49246"
    />
    <g
      id="g3284"
      style="fill: #ffd189; fill-opacity: 1;"
      transform="translate(-15.607067,3.3443722)"
    >
      <g
        id="g3272"
        style="fill: #ffd189; fill-opacity: 1;"
        transform="translate(189.51438,3.3443714)"
      >
        <rect
          height="112.59386"
          id="rect3274"
          ry="0"
          style="fill: #ffd189; fill-opacity: 1;"
          width="162.7594"
          x="152.72629"
          y="206.23622"
        />
        <rect
          height="150.49673"
          id="rect3276"
          ry="30.099342"
          style="fill: #ffd189; fill-opacity: 1;"
          width="153.84108"
          x="294.30466"
          y="187.28477"
        />
      </g>
      <g
        id="g3278"
        style="fill: #ffd189; fill-opacity: 1;"
        transform="matrix(-1,0,0,-1,595.29807,528.41065)"
      >
        <rect
          height="112.59386"
          id="rect3280"
          ry="0"
          style="fill: #ffd189; fill-opacity: 1;"
          width="162.7594"
          x="152.72629"
          y="206.23622"
        />
        <rect
          height="150.49673"
          id="rect3282"
          ry="30.099342"
          style="fill: #ffd189; fill-opacity: 1;"
          width="153.84108"
          x="294.30466"
          y="187.28477"
        />
      </g>
    </g>
  </g>
  <g
    transform="translate(-68.002197,-193.97351)"
  >
    <rect
      class="is-red"
      height="150.48511"
      ry="0"
      width="35"
      x="160"
      y="193.97354"
    />
  </g>
  <rect
    class="is-orange"
    height="150.48511"
    ry="0"
    width="35"
    x="155"
    y="0"
  />
  <rect
    class="is-violet"
    height="112.64787"
    ry="0"
    width="35"
    x="245"
    y="18.901487"
  />
  <rect
    class="is-black"
    height="112.64787"
    ry="0"
    width="35"
    x="330"
    y="18.901487"
  />
  <rect
    class="is-brown"
    height="150.48511"
    ry="0"
    width="35"
    x="447.96912"
    y="0.0264872"
  />
</svg>
`;
